zlib: switch to git source

DragonBluep opened the pull request:
Ref:
https://github.com/openwrt/openwrt/blob/08ebb9e91401416e8c8fa0519ab431d7df4c2b2c/tools/zlib/Makefile#L14-L16

Ansuel commented:
zlib is not required for tar and zstd right?

If that's the case I would switch to git source so we can pack the thing ourself (aka less space for the mirror)

DragonBluep commented:
> zlib is not required for tar and zstd right?

We are using tar and zstd from host machine so this change won't affect them.

> 
> If that's the case I would switch to git source so we can pack the thing ourself (aka less space for the mirror)

I don't like clone git because it will download more data and it seems that we can not fetch cached tarballs from openwrt mirror server.

The more important thing is that I just found there are some bugs on package source packer. **The downloader can not generate unique tar ball.** With the following patch:
```
diff --git a/package/libs/zlib/Makefile b/package/libs/zlib/Makefile
index 29183f34dd..efd35716f0 100644
--- a/package/libs/zlib/Makefile
+++ b/package/libs/zlib/Makefile
@@ -11,12 +11,13 @@ PKG_NAME:=zlib
 PKG_VERSION:=1.3.1
 PKG_RELEASE:=1
 
-PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
-PKG_SOURCE_URL:=https://github.com/madler/zlib/releases/download/v$(PKG_VERSION)
-PKG_HASH:=9a93b2b7dfdac77ceba5a558a580e74667dd6fede4585b91eefb60f03b72df23
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/madler/zlib
+PKG_SOURCE_VERSION:=51b7f2abdade71cd9bb0e7a373ef2610ec6f9daf
+# PKG_MIRROR_HASH:=
 
 PKG_LICENSE:=Zlib
-PKG_LICENSE_FILES:=README
+PKG_LICENSE_FILES:=LICENSE
 PKG_CPE_ID:=cpe:/a:gnu:zlib
 
 PKG_CONFIG_DEPENDS:= CONFIG_ZLIB_OPTIMIZE_SPEED

```
On my local WSL Debian build system:
```
db@WSL:~/owrt$ make package/zlib/download V=s
Collecting package info: done
Collecting target info: done
WARNING: Makefile 'package/feeds/packages/gst1-plugins-base/Makefile' has a dependency on 'libmesa', which does not exist
WARNING: Makefile 'package/feeds/packages/gst1-plugins-base/Makefile' has a dependency on 'libwayland', which does not exist
WARNING: Makefile 'package/feeds/packages/gst1-plugins-base/Makefile' has a dependency on 'libgraphene', which does not exist
make[2]: Entering directory '/home/db/owrt/scripts/config'
make[2]: 'conf' is up to date.
make[2]: Leaving directory '/home/db/owrt/scripts/config'
time: target/linux/prereq#0.67#0.04#0.98
WARNING: your configuration is out of sync. Please run make menuconfig, oldconfig or defconfig!
make[1]: Entering directory '/home/db/owrt'
make[2]: Entering directory '/home/db/owrt/package/libs/zlib'
mkdir -p /home/db/owrt/dl
SHELL= flock /home/db/owrt/tmp/.zlib-1.3.1.tar.zst.flock -c '    /home/db/owrt/scripts/dl_github_archive.py --dl-dir="/home/db/owrt/dl" --url="https://github.com/madler/zlib" --version="51b7f2abdade71cd9bb0e7a373ef2610ec6f9daf" --subdir="zlib-1.3.1" --source="zlib-1.3.1.tar.zst" --hash="x" --submodules  || (      echo "Checking out files from the git repository..."; mkdir -p /home/db/owrt/tmp/dl && cd /home/db/owrt/tmp/dl && rm -rf zlib-1.3.1 && [ \! -d zlib-1.3.1 ] && git clone  https://github.com/madler/zlib zlib-1.3.1 && (cd zlib-1.3.1 && git checkout 51b7f2abdade71cd9bb0e7a373ef2610ec6f9daf) && export TAR_TIMESTAMP=`cd zlib-1.3.1 && git log -1 --format='\''@%ct'\''` && echo "Generating formal git archive (apply .gitattributes rules)" && (cd zlib-1.3.1 && git config core.abbrev 8 && git archive --format=tar HEAD --output=../zlib-1.3.1.tar.git) && tar --ignore-failed-read -C zlib-1.3.1 -f zlib-1.3.1.tar.git -r .git .gitmodules 2>/dev/null && rm -rf zlib-1.3.1 && mkdir zlib-1.3.1 && tar -C zlib-1.3.1 -xf zlib-1.3.1.tar.git && (cd zlib-1.3.1 && git submodule update --init --recursive --  && rm -rf .git .gitmodules) && echo "Packing checkout..." &&  tar --numeric-owner --owner=0 --group=0 --mode=a-s --sort=name ${TAR_TIMESTAMP:+--mtime="$TAR_TIMESTAMP"} -c zlib-1.3.1 |  zstd -T0 --ultra -20 -c > /home/db/owrt/tmp/dl/zlib-1.3.1.tar.zst && mv /home/db/owrt/tmp/dl/zlib-1.3.1.tar.zst /home/db/owrt/dl/ && rm -rf zlib-1.3.1; );     '
zlib-1.3.1.tar.zst: Download from https://github.com/madler/zlib failed
zlib-1.3.1.tar.zst: Requires sha256sum for verification
Checking out files from the git repository...
Cloning into 'zlib-1.3.1'...
remote: Enumerating objects: 7284, done.
remote: Counting objects: 100% (151/151), done.
remote: Compressing objects: 100% (25/25), done.
remote: Total 7284 (delta 130), reused 131 (delta 126), pack-reused 7133 (from 2)
Receiving objects: 100% (7284/7284), 4.34 MiB | 1.24 MiB/s, done.
Resolving deltas: 100% (5138/5138), done.
Note: switching to '51b7f2abdade71cd9bb0e7a373ef2610ec6f9daf'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 51b7f2a zlib 1.3.1
Generating formal git archive (apply .gitattributes rules)
Packing checkout...
make[2]: Leaving directory '/home/db/owrt/package/libs/zlib'
time: package/libs/zlib/download#3.30#0.73#13.73
make[1]: Leaving directory '/home/db/owrt'


db@WSL:~/owrt$ sha256sum dl/zlib-1.3.1.tar.zst 
ef30c5a8f65247ffc604d8d598dbdeec3d5671f6f8523e5d4c083684e7327a6e  dl/zlib-1.3.1.tar.zst


db@WSL:~/owrt$ zstd -v
*** Zstandard CLI (64-bit) v1.5.4, by Yann Collet ***
stdin is a console, aborting
db@WSL:~/owrt$ tar --version
tar (GNU tar) 1.34
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
```
On GitHub codespace:
```
@DragonBluep ➜ /workspaces/openwrt (main) $ make package/zlib/download V=s
WARNING: Makefile 'package/utils/audit/Makefile' has a dependency on 'libev', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a build dependency on 'libpam', which does not exist
WARNING: Makefile 'package/boot/kexec-tools/Makefile' has a dependency on 'liblzma', which does not exist
WARNING: Makefile 'package/network/services/lldpd/Makefile' has a dependency on 'libnetsnmp', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a build dependency on 'libpam', which does not exist
make[2]: Entering directory '/workspaces/openwrt/scripts/config'
make[2]: 'conf' is up to date.
make[2]: Leaving directory '/workspaces/openwrt/scripts/config'
Checking 'python3-dev'... ok.
Checking 'python3-setuptools'... ok.
Checking 'swig'... ok.
make[1]: Entering directory '/workspaces/openwrt'
make[2]: Entering directory '/workspaces/openwrt/package/libs/zlib'
mkdir -p /workspaces/openwrt/dl
/workspaces/openwrt/scripts/dl_github_archive.py --dl-dir="/workspaces/openwrt/dl" --url="https://github.com/madler/zlib" --version="51b7f2abdade71cd9bb0e7a373ef2610ec6f9daf" --subdir="zlib-1.3.1" --source="zlib-1.3.1.tar.zst" --hash="x" --submodules  || (   echo "Checking out files from the git repository..."; mkdir -p /workspaces/openwrt/tmp/dl && cd /workspaces/openwrt/tmp/dl && rm -rf zlib-1.3.1 && [ \! -d zlib-1.3.1 ] && git clone  https://github.com/madler/zlib zlib-1.3.1 && (cd zlib-1.3.1 && git checkout 51b7f2abdade71cd9bb0e7a373ef2610ec6f9daf) && export TAR_TIMESTAMP=`cd zlib-1.3.1 && git log -1 --format='@%ct'` && echo "Generating formal git archive (apply .gitattributes rules)" && (cd zlib-1.3.1 && git config core.abbrev 8 && git archive --format=tar HEAD --output=../zlib-1.3.1.tar.git) && tar --ignore-failed-read -C zlib-1.3.1 -f zlib-1.3.1.tar.git -r .git .gitmodules 2>/dev/null && rm -rf zlib-1.3.1 && mkdir zlib-1.3.1 && tar -C zlib-1.3.1 -xf zlib-1.3.1.tar.git && (cd zlib-1.3.1 && git submodule update --init --recursive --  && rm -rf .git .gitmodules) && echo "Packing checkout..." &&    tar --numeric-owner --owner=0 --group=0 --mode=a-s --sort=name ${TAR_TIMESTAMP:+--mtime="$TAR_TIMESTAMP"} -c zlib-1.3.1 |       zstd -T0 --ultra -20 -c > /workspaces/openwrt/tmp/dl/zlib-1.3.1.tar.zst && mv /workspaces/openwrt/tmp/dl/zlib-1.3.1.tar.zst /workspaces/openwrt/dl/ && rm -rf zlib-1.3.1; );     
zlib-1.3.1.tar.zst: Download from https://github.com/madler/zlib failed
zlib-1.3.1.tar.zst: Requires sha256sum for verification
Checking out files from the git repository...
Cloning into 'zlib-1.3.1'...
remote: Enumerating objects: 7284, done.
remote: Counting objects: 100% (151/151), done.
remote: Compressing objects: 100% (25/25), done.
remote: Total 7284 (delta 130), reused 131 (delta 126), pack-reused 7133 (from 2)
Receiving objects: 100% (7284/7284), 4.34 MiB | 18.01 MiB/s, done.
Resolving deltas: 100% (5138/5138), done.
Note: switching to '51b7f2abdade71cd9bb0e7a373ef2610ec6f9daf'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 51b7f2a zlib 1.3.1
Generating formal git archive (apply .gitattributes rules)
Packing checkout...
make[2]: Leaving directory '/workspaces/openwrt/package/libs/zlib'
time: package/libs/zlib/download#1.83#0.49#3.62
make[1]: Leaving directory '/workspaces/openwrt'


@DragonBluep ➜ /workspaces/openwrt (main) $ sha256sum dl/zlib-1.3.1.tar.zst 
0fcbe98f70f530e36cb71150908ce20cd90ad93fb6e41a348722949ed16395e6  dl/zlib-1.3.1.tar.zst


@DragonBluep ➜ /workspaces/openwrt (main) $ zstd -v
*** Zstandard CLI (64-bit) v1.5.5, by Yann Collet ***
stdin is a console, aborting
@DragonBluep ➜ /workspaces/openwrt (main) $ tar --version
tar (GNU tar) 1.30
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
```
As you can see. For the same package, the hash values are different.

Ping @hauke @robimarko @ynezz

robimarko commented:
@DragonBluep Was tools/zstd built before that?

DragonBluep commented:
> @DragonBluep Was tools/zstd built before that?

The local Debian host has a built tools/zstd and the GitHub codespace is a new fork.

I cloned a new repo to my local Debian host `~/openwrt`, got another hash:
```
db@WSL:~/openwrt$ make package/zlib/download V=s
WARNING: Makefile 'package/utils/audit/Makefile' has a dependency on 'libev', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a build dependency on 'libpam', which does not exist
WARNING: Makefile 'package/boot/kexec-tools/Makefile' has a dependency on 'liblzma', which does not exist
WARNING: Makefile 'package/network/services/lldpd/Makefile' has a dependency on 'libnetsnmp', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a build dependency on 'libpam', which does not exist
make[2]: Entering directory '/home/db/openwrt/scripts/config'
make[2]: 'conf' is up to date.
make[2]: Leaving directory '/home/db/openwrt/scripts/config'
make[1]: Entering directory '/home/db/openwrt'
make[2]: Entering directory '/home/db/openwrt/package/libs/zlib'
mkdir -p /home/db/openwrt/dl
/home/db/openwrt/scripts/dl_github_archive.py --dl-dir="/home/db/openwrt/dl" --url="https://github.com/madler/zlib" --version="51b7f2abdade71cd9bb0e7a373ef2610ec6f9daf" --subdir="zlib-1.3.1" --source="zlib-1.3.1.tar.zst" --hash="x" --submodules  || (         echo "Checking out files from the git repository..."; mkdir -p /home/db/openwrt/tmp/dl && cd /home/db/openwrt/tmp/dl && rm -rf zlib-1.3.1 && [ \! -d zlib-1.3.1 ] && git clone  https://github.com/madler/zlib zlib-1.3.1 && (cd zlib-1.3.1 && git checkout 51b7f2abdade71cd9bb0e7a373ef2610ec6f9daf) && export TAR_TIMESTAMP=`cd zlib-1.3.1 && git log -1 --format='@%ct'` && echo "Generating formal git archive (apply .gitattributes rules)" && (cd zlib-1.3.1 && git config core.abbrev 8 && git archive --format=tar HEAD --output=../zlib-1.3.1.tar.git) && tar --ignore-failed-read -C zlib-1.3.1 -f zlib-1.3.1.tar.git -r .git .gitmodules 2>/dev/null && rm -rf zlib-1.3.1 && mkdir zlib-1.3.1 && tar -C zlib-1.3.1 -xf zlib-1.3.1.tar.git && (cd zlib-1.3.1 && git submodule update --init --recursive --  && rm -rf .git .gitmodules) && echo "Packing checkout..." &&  tar --numeric-owner --owner=0 --group=0 --mode=a-s --sort=name ${TAR_TIMESTAMP:+--mtime="$TAR_TIMESTAMP"} -c zlib-1.3.1 |       zstd -T0 --ultra -20 -c > /home/db/openwrt/tmp/dl/zlib-1.3.1.tar.zst && mv /home/db/openwrt/tmp/dl/zlib-1.3.1.tar.zst /home/db/openwrt/dl/ && rm -rf zlib-1.3.1; );     
zlib-1.3.1.tar.zst: Download from https://github.com/madler/zlib failed
zlib-1.3.1.tar.zst: Requires sha256sum for verification
Checking out files from the git repository...
Cloning into 'zlib-1.3.1'...
remote: Enumerating objects: 7284, done.
remote: Counting objects: 100% (151/151), done.
remote: Compressing objects: 100% (25/25), done.
remote: Total 7284 (delta 130), reused 131 (delta 126), pack-reused 7133 (from 2)
Receiving objects: 100% (7284/7284), 4.34 MiB | 505.00 KiB/s, done.
Resolving deltas: 100% (5138/5138), done.
Note: switching to '51b7f2abdade71cd9bb0e7a373ef2610ec6f9daf'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 51b7f2a zlib 1.3.1
Generating formal git archive (apply .gitattributes rules)
Packing checkout...
make[2]: Leaving directory '/home/db/openwrt/package/libs/zlib'
time: package/libs/zlib/download#3.31#0.91#15.17
make[1]: Leaving directory '/home/db/openwrt'


db@WSL:~/openwrt$ sha256sum dl/zlib-1.3.1.tar.zst 
414ce9c1a0675328c1169dd2d32e2aa724a9d14ade1c0b864943fd0f9ae324c2  dl/zlib-1.3.1.tar.zst
```

DragonBluep committed:
tools: zlib: switch to git source

Switch to the git repository source so that we can use zstd
compression algorithm to create smaller package tarball. This
patch also corrected the license file name[1].

[1] https://github.com/madler/zlib/commit/352cb28d12baf02863ff5d4d96be0587ced419a1

Suggested-by: Christian Marangi <ansuelsmth@gmail.com>
Signed-off-by: Shiji Yang <yangshiji66@qq.com>
Link: https://github.com/openwrt/openwrt/pull/17880
Signed-off-by: Robert Marko <robimarko@gmail.com>

DragonBluep committed:
zlib: switch to git source

Switch to the git repository source so that we can use zstd
compression algorithm to create smaller package tarball. This
patch also corrected the license file name[1] and the CPE ID[2].

[1] https://github.com/madler/zlib/commit/352cb28d12baf02863ff5d4d96be0587ced419a1
[2] https://nvd.nist.gov/products/cpe/detail/95C64A3E-A897-4D55-B74A-D2285440D164

Suggested-by: Christian Marangi <ansuelsmth@gmail.com>
Signed-off-by: Shiji Yang <yangshiji66@qq.com>
Link: https://github.com/openwrt/openwrt/pull/17880
Signed-off-by: Robert Marko <robimarko@gmail.com>

DragonBluep commented:
@Ansuel Thank you for your detailed explanation! I've switched the zlib package source to the Git repository.

Neustradamus commented:
To follow this important PR :)

robimarko commented:
Thanks! Rebased on top of main and merged!
